### ISSUE TITLE
Class has two prototypes

### ISSUE THREAD

**Issue by b5f25af5-34f1-48de-a471-e4245c39a3bd** [python/cpython#86923]
BPO | [42757](https://bugs.python.org/issue42757)
--- | :---
Nosy | @SinTh0r4s

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2020-12-27.18:01:57.552>
labels = ['type-feature', '3.7']
title = 'Class has two prototypes'
updated_at = <Date 2020-12-27.18:01:57.552>
user = 'https://github.com/SinTh0r4s'
```

bugs.python.org fields:
```python
activity = <Date 2020-12-27.18:01:57.552>
actor = 'SinTh0r4s'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = []
creation = <Date 2020-12-27.18:01:57.552>
creator = 'SinTh0r4s'
dependencies = []
files = []
hgrepos = []
issue_num = 42757
keywords = []
message_count = 1.0
messages = ['383856']
nosy_count = 1.0
nosy_names = ['SinTh0r4s']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue42757'
versions = ['Python 3.7']
```

</p></details>

**Comment by b5f25af5-34f1-48de-a471-e4245c39a3bd** [comment]
Greetings,

tl;dr Depending on how a class is imported it will have a differenct class name and therefore prototype. I ran into this issue when i used a class variable to implement an event bus.

To reproduce you need a module and a submodule. The example creates two child classes that and collects their references for further use in a class/static variable in the parent. Depending on the import description this might be broken. Here, the code:

```
module
  |--main.py
  |--child.py
  |--submodule
       |-parent.py
       |-__init.py__
```

parent.py:

```
    class Parent:

        listeners = []

        def __init__(self):
            Parent.listeners.append(self)
            print("Parent: " + str(len(Parent.listeners)))
```


child.py

```py
    from submodule.parent import Parent

    class Child(Parent):

        def __init__(self):
            super(Child, self).__init__()
            print("Child: " + str(Parent.listeners))
```


main.py

```py
    from module.child import Child
    from module.submodule.parent import Parent

    class ChildLocal(Parent):

        def __init__(self):
            super(ChildLocal, self).__init__()
            print("ChildLocal: " + str(Parent.listeners))


if __name__ == '__main__':
    child= Child()
    local = ChildLocal()
```

Result:

```
    Parent: 1
    Child: [\<module.submodule.Child object at 0x0000020F51EE2DC8\>]
    Parent: 1
    ChildLocal: [\<__main__.ChildLocal object at 0x0000020F51EE2E48\>]
```

Here, the parent object is imported in two different ways and two separate parental classes are created. Hence, each parent class collects only one child

If Parent is importet in child.py as 'from module.submodule.parent import Parent" (difference is the module.) the result looks as follows:

```
    Parent: 1
    Child: [\<module.submodule.Child object at 0x00000182B0982DC8\>]
    Parent: 2
    ChildLocal: [\<module.submodule.Child object at 0x00000182B0982DC8\>, \<__main__.ChildLocal object at 0x00000182B0982E48\>]
```

Here, both children are registered in the same parent class as intenden

In conclusion, there is a nasty (to debug) difference between a relative import path and a project root path. I am quite certain this behaviour is as intenden, but on the little chance it isn't i wanted to bring this to attentions, since it is a nightmare to find. Is it intended?

Cheers,
Sin

**Comment by methane** [comment]
This behavior is intentional.

It is well known that having the same directory accessible through two different import paths can be a source of subtle problems, but I do not have an idea for fixing it without breaking backward compatibility.
